src/BinFile/WasmParser.fs

```diff
@@ -0,0 +1,367 @@
+  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+  SOFTWARE.
+*)
+
+module internal B2R2.BinFile.Wasm.Parser
+
+open B2R2
+open B2R2.BinFile
+open B2R2.BinFile.Wasm.Section
+open System
+
+let sectionIdToName (secId: SectionId) (off: int) =
+  match secId with
+  | SectionId.Custom -> String.Format ("custom_{0:x}", off)
+  | SectionId.Type -> "type"
+  | SectionId.Import -> "import"
+  | SectionId.Function -> "function"
+  | SectionId.Table -> "table"
+  | SectionId.Memory -> "memory"
+  | SectionId.Global -> "global"
+  | SectionId.Export -> "export"
+  | SectionId.Start -> "start"
+  | SectionId.Element -> "element"
+  | SectionId.Code -> "code"
+  | SectionId.Data -> "data"
+  | _ -> ""
+
+let private summerizeSections (reader: BinReader) offset =
+  let rec loop (acc: _ list) no =
+    if reader.IsOutOfRange no then
+      acc
+    else
+      let id, size, len = peekSectionHeader reader no
+      let headerSize = len + 1
+      let no' = no + headerSize + int size
+      let summary = {
+        Id = id
+        Name = sectionIdToName id no
+        Offset = no
+        HeaderSize = uint32 headerSize
+        ContentsSize = size
+      }
+      loop (acc @ [summary]) no'
+  loop [] offset
+
+let private idLtId id1 id2 =
+  let id1' = LanguagePrimitives.EnumToValue id1
+  let id2' = LanguagePrimitives.EnumToValue id2
+  id1' < id2'
+
+let private peekSecSummPair (secsSumm: SectionSummary list) =
+  let sec1 = List.head secsSumm
+  let secsSumm' = List.tail secsSumm
+  let sec2 = List.tryHead secsSumm'
+  sec1, sec2, secsSumm'
+
+let validateSectionsOrder secsSummary =
+  let rec validationLoop secsSumm isValid =
+    if List.length secsSumm = 0
+    then isValid
+    else
+    let sec1, sec2, secsSumm' = peekSecSummPair secsSumm
+    match sec2 with
+      | Some sec ->
+        let id1 = sec1.Id
+        let id2 = sec.Id
+        let isValid' =
+          if id1 = SectionId.Custom || id2 = SectionId.Custom
+          then true
+          else idLtId id1 id2
+        if not isValid' then isValid'
+        else validationLoop secsSumm' isValid'
+      | None -> isValid
+  validationLoop secsSummary true
+
+let updateSection wm id updateRec parseSec (secsSumm: SectionSummary list) =
+  let secSumm =
+    secsSumm
+    |> List.filter (fun sm -> sm.Id = id)
+    |> List.tryHead
+  match secSumm with
+    | Some sm ->
+      let secsSummary' =
+        secsSumm |> List.except [sm]
+      let sec =
+        parseSec wm.BinReader sm.Offset
+      (updateRec wm sec), secsSummary'
+    | None -> wm, secsSumm
+
+let updateCustomSection wasmModule secsSummary =
+  let ur wm sec =
+    {
+      wm with
+        CustomSections = wm.CustomSections @ [ sec ]
+    }
```

Comment:
This can be in one liner: which consumes less space and seems concise to me. Similarly for the below constructs too.